### PR TITLE
Update record sanity check

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #2686

## What is the purpose of the change

> Add sanity check in `updateRecord` to compare record time and height against `ctx`

## Brief Changelog
 
  - *Created new error type*
  - *Add err handle into `updateRecord` & `UpdateRecords`*
  - *Adjusted test*

### CODE REVIEW

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:1030]: Can we also add a test case for having block height greater than the last record's time and checking the error?

**Review by mattverse — COMMENTED** [review]
Looks nice! I think we need a test case for checking error on block height and then should be good to merge! 

Also curious if the issue / question you asked on the original issue is resolved or if its something that you need further answer on

File: x/twap/logic_test.go
Comment (hieuvubk) [reply x/twap/logic_test.go:1032]: Yes of course

File: x/twap/types/errors.go
Comment (alexanderbez) [inline x/twap/types/errors.go:74]:
```suggestion
	return fmt.Sprintf("failed to update the record, the context time must be greater than record time; record: block %d at %s, actual: block %d at %s", e.RecordBlockHeight, e.RecordTime, e.ActualBlockHeight, e.ActualTime)
```

**Review by alexanderbez — APPROVED** [review]

File: x/twap/types/errors.go
Comment (hieuvubk) [reply x/twap/types/errors.go]: thanks you

**Review by alexanderbez — APPROVED** [review]

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:1257]: Can we use the `WithBlockTime` and `WithBlockHeight` API instead?

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:1013]:
```suggestion
		"new record can't be inserted prior to the last record's update time": {
```

File: x/twap/types/errors.go
Comment (mattverse) [inline x/twap/types/errors.go:66]: I'm wondering if we should change this to `InvalidUpdateRecordError` or something similar, since this error does not only happen when it has a invalid record **time** but it also includes invlaid record height as well. Lmk what you think

**Comment by hieuvubk** [comment]
nice thank u

File: x/twap/logic.go
Comment (mattverse) [inline x/twap/logic.go:163]: Can we add test cases to verify all these logic gates?
Right now only one test case has been added where we test if record.Height ctx.BlockHeight, but it'd be great if we could have additional tests for
1. same test but with record
2. same test as above but with P1ArithmeticTwapAccumulator ==0 and P0ArithmeticTwapAccumulator ==0
3. same test as above but with P1ArithmeticTwapAccumulator ==0 and P0ArithmeticTwapAccumulator !=0

**Review by mattverse — COMMENTED** [review]
Nice work, logic lgtm, requesting for additional test cases before we can merge!

File: x/twap/logic.go
Comment (hieuvubk) [reply x/twap/logic.go:162]: Its a case when creating new pool. New twap records will be created and updated immediately. I limit this by adding the condition `ArithmeticTwapAccumulator` must be zero (init twap record). But I should split it up

Comment (mattverse) [reply x/twap/logic.go]: sweet, can we add this context as inline comment?

File: x/twap/logic.go
Comment (mattverse) [inline x/twap/logic.go:169]: we should be using `		return types.TwapRecord{}, types.InvalidUpdateRecordError{` here as well

**Review by mattverse — APPROVED** [review]
One comment regarding returned error type. Otherwise lgtm

**Comment by mattverse** [comment]
Sweet! Let's merge this after updating CHANGELOG please @hieuvubk

**Comment by hieuvubk** [comment]
@mattverse hey sr I forgot about this pr :((. Changelog updated

**Review by czarcas7ic — APPROVED** [review]

**Pull request merged**
